kafka-client/src/main/java/dev/responsive/kafka/internal/db/MongoSessionFlushManager.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2024 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.db;
+
+import com.mongodb.MongoBulkWriteException;
+import com.mongodb.MongoException;
+import com.mongodb.bulk.WriteConcernError;
+import com.mongodb.client.MongoCollection;
+import dev.responsive.kafka.internal.db.mongo.MongoSessionTable;
+import dev.responsive.kafka.internal.db.mongo.MongoWriter;
+import dev.responsive.kafka.internal.db.mongo.SessionDoc;
+import dev.responsive.kafka.internal.db.partitioning.Segmenter.SegmentPartition;
+import dev.responsive.kafka.internal.db.partitioning.SessionSegmentPartitioner;
+import dev.responsive.kafka.internal.db.partitioning.TablePartitioner;
+import dev.responsive.kafka.internal.stores.RemoteWriteResult;
+import dev.responsive.kafka.internal.utils.SessionKey;
+import java.util.function.Function;
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+public class MongoSessionFlushManager extends SessionFlushManager {
+
+  private final String logPrefix;
+  private final Logger log;
+
+  private final MongoSessionTable table;
```

Comment:
Is this class basically the same as the MongoWindowFlushManager except for the type of this table (and the partitioner)? Just wondering, did you look into whether it's possible to consolidate them by parametrizing somehow?

My main concern with all this is: how easy would it be to make a change in one class but not the other, and accidentally screw up one of the implementations. It's not the end of the world to have repeated code, but what's the risk of code which should be identical getting out of sync?

Reply:
You're right, it is almost entirely a copy paste of the MongoWindowFlushManager class, which itself is a relatively close replica of the MongoKVFlushManager. My plan was to refactor these 3 into 1 separately in another PR so as to keep this one entirely devoid of any changes impacting other stores (KV / Window).